File: day2/day2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Opcode string

const (
	OpForward Opcode = "forward"
	OpDown    Opcode = "down"
	OpUp      Opcode = "up"
)

type Operation struct {
	OpCode  Opcode
	Operand int
}

func readInput(filePath string) ([]Operation, error) {
	file, err := os.Open(filePath)

	if err != nil {
		return nil, err
	}

	scanner := bufio.NewScanner(file)

	result := []Operation{}

	var (
		opBuffer []string
		op       Operation
	)

	for scanner.Scan() {
		opBuffer = strings.Split(scanner.Text(), " ")
		if len(opBuffer) != 2 {
			continue
		}

		op = Operation{}
		switch opBuffer[0] {
		case string(OpUp):
			op.OpCode = OpUp
		case string(OpDown):
			op.OpCode = OpDown
		case string(OpForward):
			op.OpCode = OpForward
		default:
			continue
		}

		op.Operand, err = strconv.Atoi(opBuffer[1])
		if err != nil {
			return nil, fmt.Errorf("cannot convert operand: %w", err)
		}

		result = append(result, op)
	}

	return result, nil
}

func part1(input []Operation) int {
	var (
		x int
		y int
	)

	for _, op := range input {
		switch op.OpCode {
		case OpForward:
			x += op.Operand
		case OpDown:
			y += op.Operand
		case OpUp:
			y -= op.Operand
		}
	}

	return x * y
}

func part2(input []Operation) int {
	var (
		x   int
		y   int
		aim int
	)

	for _, op := range input {
		switch op.OpCode {
		case OpForward:
			x += op.Operand
			y += op.Operand * aim
		case OpDown:
			aim += op.Operand
		case OpUp:
			aim -= op.Operand
		}
	}

	return x * y
}

func main() {
	input, err := readInput("../inputs/day2.txt")
	if err != nil {
		panic(err)
	}

	fmt.Printf("Part 1: %d\n", part1(input))
	fmt.Printf("Part 2: %d\n", part2(input))
}

File: day2/day2.rs
use std::fs;
use std::vec::Vec;

#[derive(Debug)]
enum OpCode {
    None,
    Forward,
    Down,
    Up,
}

#[derive(Debug)]
struct Op {
    code: OpCode,
    operand: i32,
}

impl Op {
    pub fn new(code: &str, operand: &str) -> Self {
        let code = match code {
            "forward" => OpCode::Forward,
            "down" => OpCode::Down,
            "up" => OpCode::Up,
            _ => OpCode::None,
        };

        let operand = operand.parse().unwrap();

        Op{code, operand} 
    }
}

fn main() -> std::io::Result<()> {
    let input_str = fs::read_to_string("../inputs/day2.txt")?;

    let input: Vec<Op> = input_str
        .lines()
        .map(|line| line.split_ascii_whitespace().collect())
        .filter(|op: &Vec<&str>| op.len() == 2)
        .map(|op: Vec<&str>| Op::new(op[0], op[1]))
        .collect();
        

    println!("Part 1: {:?}", part1(&input));
    println!("Part 2: {:?}", part2(&input));
    
    Ok(())
}

fn part1(input: &[Op]) -> i32 {
    let (mut x, mut y) = (0i32, 0i32);
    input.iter().for_each(|op| match op.code {
        OpCode::Forward => x += op.operand,
        OpCode::Down => y += op.operand,
        OpCode::Up => y -= op.operand,
        _ => {},
    });

    x*y
}


fn part2(input: &Vec<Op>) -> i32 {
    let (mut x, mut y, mut aim) = (0i32, 0i32, 0i32);
    input.iter().for_each(|op| match op.code {
        OpCode::Forward =>  { x += op.operand; y += aim*op.operand },
        OpCode::Down => aim += op.operand,
        OpCode::Up => aim -= op.operand,
        _ => {},
    });

    x*y
}